### PR TITLE
Added FlexibleReLU activation layer

### PR DESCRIPTION
**Pull request by aarushgupta**
Added FlexibleRelu activation layer as mentioned in this [paper](https://arxiv.org/abs/1706.08098). This activation function is particularly useful for adding a trainable bias term to each output value from a convolutional filter.

### CODE REVIEW

**Comment by zoq** [comment]
Can you add a test for the FReLU layer?

**Comment by aarushgupta** [comment]
Sure

File: src/mlpack/methods/ann/layer/flexible_relu.hpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * @code
+ * @article{
+ *  author  = {Suo Qiu, Xiangmin Xu and Bolun Cai},
+ *  title   = {FReLU: Flexible Rectified Linear Units for Improving
+ *             Convolutional Neural Networks}
+ *  journal = {arxiv preprint},
+ *  year    = {2018}
+ * }
+ * @endcode
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/flexible_relu.hpp]: Can you add this to the documentation for the class itself also please?  A user may not necessarily find it here.

Comment (aarushgupta) [reply src/mlpack/methods/ann/layer/flexible_relu.hpp]: Sure.

**Comment by aarushgupta** [comment]
@zoq @rcurtin I have made the changes. Please review it once.

**Comment by zoq** [comment]
Do you think we could reuse the existing relu function and just add ``alpha``?

**Comment by aarushgupta** [comment]
In that case, how would we differentiate between non-trainable(and zero) and trainable alpha?

**Comment by zoq** [comment]
Not sure, I get the issue, I was thinking instead of using the ``Fn()`` from the class, we could use the ``RectifierFunction`` inside the forward pass and add ``alpha``. Maybe I missed something?

**Comment by aarushgupta** [comment]
Yeah, I think we can do that. We won't require flexible_relu.hpp then?

**Comment by aarushgupta** [comment]
Got it. I will remove it and do the necessary changes.

**Comment by aarushgupta** [comment]
@zoq The function RectifierFunction::Fn() does not return a value for matrices and tensors whereas it returns a double type value for single double type values. How should I handle this? Maybe we could check the shape of the input and output variables before calling RectifierFunction::Fn(input, output)?

**Comment by zoq** [comment]
I see, I guess as simple solution is to do something like ``Fn(input, output); output += alpha;``but I'm not sure if this is significantly slower as doing everything in a single line. So I guess what we should do here, is to go the direct way. But I think we can still simplify the code here, e.g. remove the extra methods and perform the operation inside the Forward function itself. Let me know what you think.

**Comment by aarushgupta** [comment]
Sorry for the late reply. I guess in doing so, we will have to write separate cases for each of tensors, matrices, and scalers which is already there in the flexible_relu.hpp file.

**Comment by zoq** [comment]
Just for the matrix object, the scalar method is never used, it's private anyway.

**Comment by Manthan-R-Sheth** [comment]
@zoq 
@aarushgupta any updates on this? I can take this up if its fine.
Thanks.

**Comment by aarushgupta** [comment]
Sure!

**Comment by sourabhvarshney111** [comment]
@Manthan-R-Sheth If you have rebased this PR, can @aarushgupta please close this pr?

**Comment by Manthan-R-Sheth** [comment]
I am fine with whatever @aarushgupta , @zoq and @rcurtin have to say in this.

**Comment by zoq** [comment]
I'll close this one once the other PR is merged.

**Comment by zoq** [comment]
@aarushgupta thanks again, merged with #1341.

**Pull request closed without merging**